### PR TITLE
resource: cache resource status information to reduce broker load

### PR DESCRIPTION
**Pull request by garlick**
Problem: on a large system, frequent use of `flux resource list` or `flux resource status` triggers significant broker CPU consumption.  Commands like `flux top` or `watch flux resource list` are not unreasonable things to do, yet they are costly.

As noted in #6009, we repeatedly calculate the same resource sets, even when they don't change.

Add a simple cache.

### CODE REVIEW

**Comment by garlick** [comment]
Pushed a little tweak just now but basically this is ready for review.

File: src/modules/resource/status.c [hunk 1]
```diff
@@ -442,6 +463,23 @@ static void sched_status_cb (flux_t *h,
...
+ */
+static void reslog_cb (struct reslog *reslog,
+                       const char *name,
+                       json_t *context,
+                       void *arg)
+{
+    struct status *status = arg;
+
+    invalidate_cache (&status->cache);
+}
```
Comment (grondo) [inline src/modules/resource/status.c:481]: In perf tracing, creating the `all` resource set was the most expensive operation, and `R_all` should only be invalidated by the first two events in the list above. Would it add too much complexity to save `R_all` unless a `resource-define` or `resource-update` event is seen? (you could perhaps add an optional event name parameter to `invalidate_cache`?)

On a very large system I fear the `online` `offline` `drain` and `undrain` events will be so frequent that we won't see as much benefit from the caching as we might otherwise.

File: src/modules/resource/status.c [hunk 2]
```diff
@@ -8,6 +8,12 @@
  * SPDX-License-Identifier: LGPL-3.0
  \************************************************************/
 
+/* status.c - handle resource.status and resource.sched-status RPCs
+ *
+ * Although this portion of the resource module is initialized on all
+ * ranks, the RPCs are denied if requested on rank > 0.
+ */
+
```
Comment (grondo) [inline src/modules/resource/status.c:16]: Not critical, but it seems like this comment could go in a separate commit? (The comment does make me wonder why the status portion of the `resource` module is initialized on all ranks when RPCs are denied on those ranks other than 0, though)

**Review by grondo — APPROVED** [review]
Nice! This LGTM! One suggestion inline. If you feel strongly the current approach is good, though feel free to merge as it is a big improvement.

File: src/modules/resource/status.c (hunk 1 above)
Comment (garlick) [reply src/modules/resource/status.c:481]: Sure, I'll do that.

File: src/modules/resource/status.c (hunk 2 above)
Comment (garlick) [reply src/modules/resource/status.c:16]: Sure, will do.

> The comment does make me wonder why the status portion of the resource module is initialized on all ranks when RPCs are denied on those ranks other than 0, though)

Just to deny the RPCs with a better message than "service not implemented' I think?  Hmm, maybe not worth it.  :shrug:

Comment (grondo) [reply src/modules/resource/status.c:16]: Eh, not really worth addressing now I'd say

**Comment by garlick** [comment]
I reworked the caching so that `R_all`  and the rlist are not invalidated by online/offline/drain/undrain.

One thing I'm sort of not happy with is the extra rlist copy here.  I had to add it because before, we marked nodes down in the primary rlist and then just had to copy it once with `rlist_copy_down()`.  Now in order to not cache down info in the main rlist, I'm making a copy then marking things up and down in the copy, then calling `rlist_copy_down()`.  Boy I feel dumb even typing this because I'm sure there's a better way.  I'll have a look tomorrow.
```c
static json_t *get_down (struct status *status, const struct rlist *rl)
{
    if (!status->cache.R_down) {
        struct rlist *cpy;
        struct rlist *rdown = NULL;
        struct idset *drain = NULL;
        const struct idset *down = monitor_get_down (status->ctx->monitor);

        if ((cpy = rlist_copy_empty (rl))
            && rlist_mark_up (cpy, "all") >= 0
            && (drain = drain_get (status->ctx->drain))
            && mark_down (cpy, down) == 0
            && mark_down (cpy, drain) == 0
            && (rdown = rlist_copy_down (cpy)))
            status->cache.R_down = rlist_to_R (rdown);
        idset_destroy (drain);
        rlist_destroy (rdown);
        rlist_destroy (cpy);
    }
    return status->cache.R_down;
}
```

**Comment by garlick** [comment]
Re-pushed a new version of that function that just modifies its rlist argument, which avoids the extra copy and should be fine for all users of `cache->rl` (comment to that effect added).

**Comment by garlick** [comment]
I'll go ahead and set MWP on this one.
